compiler/rustc_mir_build/src/builder/expr/into.rs

```diff
@@ -365,6 +365,113 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
+                        this.expr_into_dest(ptr_deref, block, val)
+                    }
+                    sym::init_box_via_move => {
+                        // `write_via_move(b, val)` becomes
+                        // ```
+                        // *transmute::<_, *mut T>(b) = val;
+                        // transmute::<_, Box<T>>(b)
+                        // ```
+                        let t = generic_args.type_at(0);
+                        let [b, val] = **args else {
+                            span_bug!(expr_span, "invalid init_box_via_move call")
+                        };
+                        let Some(b) = unpack!(block = this.as_local_operand(block, b)).place()
+                        else {
+                            span_bug!(expr_span, "invalid init_box_via_move call")
+                        };
+                        // Project to the pointer inside `b`. We have to keep `b` in scope to ensure
+                        // it gets dropped. After the first projection we can transmute which is
+                        // easier.
+                        let ty::Adt(box_adt_def, box_adt_args) =
+                            b.ty(&this.local_decls, this.tcx).ty.kind()
+                        else {
+                            span_bug!(expr_span, "invalid init_box_via_move call")
+                        };
+                        let unique_field =
+                            this.tcx.adt_def(box_adt_def.did()).non_enum_variant().fields
+                                [rustc_abi::FieldIdx::ZERO]
+                                .did;
+                        let Some(unique_def) =
+                            this.tcx.type_of(unique_field).instantiate_identity().ty_adt_def()
+                        else {
+                            span_bug!(
+                                this.tcx.def_span(unique_field),
+                                "expected Box to contain Unique"
+                            )
+                        };
+                        let unique_ty =
+                            Ty::new_adt(this.tcx, unique_def, this.tcx.mk_args(&[box_adt_args[0]]));
+                        let b_field = b.project_deeper(
+                            &[ProjectionElem::Field(rustc_abi::FieldIdx::ZERO, unique_ty)],
+                            this.tcx,
+                        );
+                        // `ptr` is `b` transmuted to `*mut T`.
+                        let ptr_ty = Ty::new_mut_ptr(this.tcx, t);
+                        let ptr = this.local_decls.push(LocalDecl::new(ptr_ty, expr_span));
+                        this.cfg.push(
+                            block,
+                            Statement::new(source_info, StatementKind::StorageLive(ptr)),
+                        );
+                        // Make sure `StorageDead` gets emitted.
+                        this.schedule_drop_storage_and_value(expr_span, this.local_scope(), ptr);
```

Comment:
Here I am completely guessing... well really for all the changes in this file I am guessing, but the drop/storage scope stuff I know even less about than the rest of this.